src/neurostatslib/base_class.py

```diff
@@ -0,0 +1,578 @@
+        pass
+
+    @abc.abstractmethod
+    def simulate(
+        self,
+        random_key: jax.random.PRNGKeyArray,
+        feed_forward_input: Union[NDArray, jnp.ndarray],
+    ):
+        """Simulate neural activity in response to a feed-forward input and recurrent activity."""
+        pass
+
+    @staticmethod
+    def _convert_to_jnp_ndarray(
+        *args: Union[NDArray, jnp.ndarray], data_type: Optional[jnp.dtype] = None
+    ) -> Tuple[jnp.ndarray, ...]:
+        """Convert provided arrays to jnp.ndarray of specified type.
+
+        Parameters
+        ----------
+        *args :
+            Input arrays to convert.
+        data_type :
+            Data type to convert to. Default is None, which means that the data-type
+            is inferred from the input.
+
+        Returns
+        -------
+        :
+            Converted arrays.
+        """
+        return tuple(jnp.asarray(arg, dtype=data_type) for arg in args)
+
+    @staticmethod
+    def _has_invalid_entry(array: jnp.ndarray) -> bool:
+        """Check if the array has nans or infs.
+
+        Parameters
+        ----------
+        array:
+            The array to be checked.
+
+        Returns
+        -------
+            True if a nan or an inf is present, False otherwise
+
+        """
+        return (jnp.isinf(array) | jnp.isnan(array)).any()
+
+    @staticmethod
+    def _check_and_convert_params(
+        params: Tuple[ArrayLike, ArrayLike], data_type: Optional[jnp.dtype] = None
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        """
+        Validate the dimensions and consistency of parameters and data.
+
+        This function checks the consistency of shapes and dimensions for model
+        parameters.
+        It ensures that the parameters and data are compatible for the model.
+
+        """
+        if not is_sequence(params):
+            raise TypeError("Initial parameters must be array-like!")
+
+        if len(params) != 2:
+            raise ValueError("Params needs to be array-like of length two.")
+
+        try:
+            params = jnp.asarray(params[0], dtype=data_type), jnp.asarray(
+                params[1], dtype=data_type
+            )
+        except (ValueError, TypeError):
+            raise TypeError(
+                "Initial parameters must be array-like of array-like objects"
+                "with numeric data-type!"
+            )
+
+        if params[0].ndim != 2:
+            raise ValueError(
+                "params[0] must be of shape (n_neurons, n_features), but"
+                f"params[0] has {params[0].ndim} dimensions!"
+            )
+        if params[1].ndim != 1:
+            raise ValueError(
+                "params[1] must be of shape (n_neurons,) but "
+                f"params[1] has {params[1].ndim} dimensions!"
+            )
+        return params
+
+    @staticmethod
+    def _check_input_dimensionality(
+        X: Optional[jnp.ndarray] = None, y: Optional[jnp.ndarray] = None
```

Comment:
this is a logic I might want to change:

1. the dims are not generic, they work for GLMs but other estimators may work differently (we may move the check in the GLM class or in a intermediate regressor class)

2. The `_check_input_dimensionality` requires at least one of the two to be non-empty. The reason I implemented it this way is because it is called with both X and y as input, only X or only y. not very elegant.

I am open to any more elegant solution

Reply:
I would wait on this because for now, any regressor will likely have a similar input structure. But when we will which to pytrees, we may want to rethink how to check the inputs and the parameters and where the checks should be located. (i.e. moving things around at this point would mean wasted effort, including changing my tests, to then redo things once again in the next PR)